File: docs/source/tutorial_code/checkpointing_reaction_partial.py
import logging

from libmuscle import Grid, Instance, Message, USES_CHECKPOINT_API
from ymmsl import Operator


def reaction() -> None:
    """A simple exponential reaction model on a 1D grid.
    """
    instance = Instance({
            Operator.F_INIT: ['initial_state'],     # list of float
            Operator.O_F: ['final_state']},         # list of float
            USES_CHECKPOINT_API)

    while instance.reuse_instance():
        # F_INIT
        t_max = instance.get_setting('t_max', 'float')
        dt = instance.get_setting('dt', 'float')
        k = instance.get_setting('k', 'float')

        msg = instance.receive('initial_state')
        U = msg.data.array.copy()

        t_cur = msg.timestamp
        t_stop = msg.timestamp + t_max
        while t_cur + dt < t_stop:
            # O_I

            # S
            U += k * U * dt
            t_cur += dt

            if instance.should_save_snapshot(t_cur):
                msg = Message(t_cur, data=[Grid(U, ['x']), t_stop])
                instance.save_snapshot(msg)

        # O_F
        instance.send('final_state', Message(t_cur, data=Grid(U, ['x'])))

        if instance.should_save_final_snapshot():
            msg = Message(t_cur)
            instance.save_final_snapshot(msg)


if __name__ == '__main__':
    logging.basicConfig()
    logging.getLogger().setLevel(logging.INFO)
    reaction()

File: docs/source/tutorial_code/checkpointing_reaction_partial.cpp
#include <cstdlib>
#include <vector>

#include <libmuscle/libmuscle.hpp>
#include <ymmsl/ymmsl.hpp>


using libmuscle::Data;
using libmuscle::DataConstRef;
using libmuscle::Instance;
using libmuscle::InstanceFlags;
using libmuscle::Message;
using ymmsl::Operator;


/** A simple exponential reaction model on a 1D grid.
 */
void reaction(int argc, char * argv[]) {
    Instance instance(argc, argv, {
            {Operator::F_INIT, {"initial_state"}},  // 1D Grid
            {Operator::O_F, {"final_state"}}},      // 1D Grid
            InstanceFlags::USES_CHECKPOINT_API);

    while (instance.reuse_instance()) {

        // F_INIT
        double t_max = instance.get_setting_as<double>("t_max");
        double dt = instance.get_setting_as<double>("dt");
        double k = instance.get_setting_as<double>("k");

        auto msg = instance.receive("initial_state");
        auto data_ptr = msg.data().elements<double>();
        std::vector<double> U(data_ptr, data_ptr + msg.data().size());

        double t_cur = msg.timestamp();
        double t_stop = msg.timestamp() + t_max;
        while (t_cur + dt < t_stop) {
            // O_I

            // S
            for (double & u : U)
                u += k * u * dt;
            t_cur += dt;

            if (instance.should_save_snapshot(t_cur)) {
                Message msg(t_cur,
                        Data::list(Data::grid(U.data(), {U.size()}, {"x"}), t_stop));
                instance.save_snapshot(msg);
            }
        }

        // O_F
        auto result = Data::grid(U.data(), {U.size()}, {"x"});
        instance.send("final_state", Message(t_cur, result));

        if (instance.should_save_final_snapshot()) {
            Message msg(t_cur);
            instance.save_final_snapshot(msg);
        }
    }
}


int main(int argc, char * argv[]) {
    reaction(argc, argv);
    return EXIT_SUCCESS;
}

